### PR TITLE
HV-1420 Add a value extraction chapter to the doc

### PR DESCRIPTION
**Pull request by gsmet**
* https://hibernate.atlassian.net/browse/HV-1420

### CODE REVIEW

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 1]
```diff
@@ -1,3 +1,324 @@
 [[chapter-valueextraction]]
 == Value extraction
 
+Value extraction is the process extracting values from a container so that
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: the process *of*

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 2]
```diff
@@ -1,3 +1,324 @@
...
+
+It is used when dealing with <<container-element-constraints,container element
+constraints>> and <<example-cascaded-validation-list,cascaded validation inside
+containers>>.
+
+[[section-valueextraction-builtinvalueextractors]]
+=== Built-in value extractors
+
+Hibernate Validator comes with built-in value extractors for the usual Java container
+types so, apart if you are using your own custom container types (or the ones
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: apart => except? Not sure about that, but I've never heard "apart if ...", always "apart from -ing".

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 3]
```diff
@@ -1,3 +1,324 @@
...
+of external libraries such as Guava's `Multimap`), you should not have to add your
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Maybe a link to Guava's website would be useful?

File: documentation/src/test/java/org/hibernate/validator/referenceguide/chapter07/valueextractor/MultimapKeyValueExtractor.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import com.google.common.collect.Multimap;
+
+//tag::include[]
+public class MultimapKeyValueExtractor
+		implements ValueExtractor<Multimap<@ExtractedValue ?, ?>> {
+
+	@Override
+	public void extractValues(Multimap<?, ?> originalValue, ValueReceiver receiver) {
+		for ( Object key : originalValue.keySet() ) {
+			receiver.keyedValue( "<multimap key>", key, key );
```
Comment (yrodiere) [inline documentation/src/test/java/org/hibernate/validator/referenceguide/chapter07/valueextractor/MultimapKeyValueExtractor.java]: I find it a bit weird that we pass the key as a "keyed value" with the key as its own key... I would have expected something like `receiver.iterableValue("<multimap key>", key)`. Maybe it's worth some additional explanation?

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 5]
```diff
@@ -1,3 +1,324 @@
...
+    for a container containing indexed values - it is used for ``List``s
+`keyedValue()`::
+    for a container containing keyed values - it is used for ``Map``s
+
+For all these methods, you need to pass a node name: it is the name included in
+the node added to the property path of the constraint violation. As mentioned
+earlier, if the node name is `null`, no node is added to the property path:
+it is be useful for pure wrapper types similar to `Optional`.
+
+The choice of the method used is important as it adds important information to the
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: "adds important information" => "adds valuable information"? You already said "important" just a few words before. Alternatively, you could explain why it's important. "adds contextual information", maybe.

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 6]
```diff
@@ -1,3 +1,324 @@
...
+.A `ValueExtractor` for `OptionalInt` marked with `@UnwrapByDefault`
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/validator/referenceguide/chapter07/nongeneric/UnwrapByDefaultOptionalIntValueExtractor.java[tags=include]
+----
+====
+
+When declaring this value extractor for `OptionalInt`, constraint annotations will
+by default concern the wrapped value:
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: You're using "concern" a lot, maybe you could replace some occurrences with "relate to" or "will be applied to".

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 7]
```diff
@@ -1,3 +1,324 @@
...
+
+Bean properties in JavaFX are typically not of simple data types like `String`
+or `int`, but are wrapped in `Property` types which allows to make them observable,
+use them for data binding etc.
+
+Thus, value extraction is required to be able to apply constraints on the
+wrapped values.
+
+The JavaFX `ObservableValue` value extractor is marked with `@UnwrapByDefault`.
+As such, the constraints hosted on the container targets the wrapped value by
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: targets => target

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 8]
```diff
@@ -1,3 +1,324 @@
...
+ * {javafxDocsUrl}[JavaFX]'s `ObservableValue` (see <<section-valueextraction-javafx>>
+   for more details).
+
+The complete list of built-in value extractors with all the details on how they
+behave can be found in the
+{bvSpecUrl}#valueextractordefinition-builtinvalueextractors[Bean Validation specification].
+
+=== Implementing a `ValueExtractor`
+
+To extract values from a custom container, one needs to implement a `ValueExtractor`.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Shouldn't you mention from the start that implementing it is not enough, and it needs to be registered afterwards?

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 9]
```diff
@@ -1,3 +1,324 @@
...
+There are several ways to register value extractors (in increasing order of
+priority):
+
+Provided by the validation engine itself::
+    See <<section-valueextraction-builtinvalueextractors>>.
+
+Via the Java service loader mechanism::
+    The file `META-INF/services/javax.validation.valueextraction.ValueExtractor`
+    must be provided, with the fully-qualified names of one or more value
+    extractor implementations as its contents.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: "each on a different line"? "comma-separated"? "separated by whitespaces"?

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 10]
```diff
@@ -1,3 +1,324 @@
...
+
+In most cases, you should not have to worry about this but, if you are overriding
+existing value extractors, you can find a detailed description of the value
+extractors resolution algorithms in the Bean Validation specification:
+
+ * for {bvSpecUrl}#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-algorithm-constraints[container element constraints],
+ * for {bvSpecUrl}#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-algorithm-cascaded[cascaded validation],
+ * and for {bvSpecUrl}http://beanvalidation.org/2.0/spec/#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-implicitunwrapping[implicit unwrapping].
+
+One important thing to have in mind is that:
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: That's two things... :p

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 1 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Fixed.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 2 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Fixed.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 3 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Done.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 5 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Yeah, let's go for contextual.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 9 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Right.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 6 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Done.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 7 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Done.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 10 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Well, no, that's one global thing :). I'll keep it that way.

File: documentation/src/main/asciidoc/ch07.asciidoc (hunk 8 above)
Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Makes sense, yes. I was expecting people to read the whole paragraph :).

**Comment by gsmet** [comment]
@yrodiere Rebased and added a fixup commit to fix your remarks.

File: documentation/src/main/asciidoc/ch07.asciidoc [hunk 11]
```diff
@@ -1,3 +1,324 @@
...
+
+=== Resolution algorithms
+
+In most cases, you should not have to worry about this but, if you are overriding
+existing value extractors, you can find a detailed description of the value
+extractors resolution algorithms in the Bean Validation specification:
+
+ * for {bvSpecUrl}#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-algorithm-constraints[container element constraints],
+ * for {bvSpecUrl}#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-algorithm-cascaded[cascaded validation],
+ * and for {bvSpecUrl}http://beanvalidation.org/2.0/spec/#constraintdeclarationvalidationprocess-validationroutine-valueextractorresolution-implicitunwrapping[implicit unwrapping].
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/ch07.asciidoc]: There's a typo here (an extra "http://beanvalidation.org/2.0/spec/")

Comment (gsmet) [inline documentation/src/main/asciidoc/ch07.asciidoc]: Good catch, rebase issue.

**Comment by gsmet** [comment]
Rebased and merged. Kinda surprised that git didn't succeed in dealing with the file renamings and the previous commit integrating Marko's blog post.

**Pull request closed without merging**